src/auth/awsrequestsigner.ts

```diff
@@ -0,0 +1,280 @@
+        Authorization: headerMap.authorizationHeader,
+        host: uri.host,
+      },
+      additionalAmzHeaders || {}
+    );
+    if (awsSecurityCredentials.token) {
+      Object.assign(headers, {
+        'x-amz-security-token': awsSecurityCredentials.token,
+      });
+    }
+    const awsSignedReq: GaxiosOptions = {
+      url,
+      method: method,
+      headers,
+    };
+
+    if (typeof requestPayload !== 'undefined') {
+      awsSignedReq.body = requestPayload;
+    }
+
+    return awsSignedReq;
+  }
+}
+
+/**
+ * Creates the HMAC-SHA256 hash of the provided message using the
+ * provided key.
+ *
+ * @param key The HMAC-SHA256 key to use.
+ * @param msg The message to hash.
+ * @return The computed hash bytes.
+ */
+async function sign(
+  crypto: Crypto,
+  key: string | ArrayBuffer,
+  msg: string
+): Promise<ArrayBuffer> {
+  return await crypto.signWithHmacSha256(key, msg);
+}
+
+/**
+ * Calculates the signature for AWS Signature Version 4.
+ * Based on:
+ * https://docs.aws.amazon.com/general/latest/gr/sigv4-calculate-signature.html
+ *
+ * @param key The AWS secret access key.
+ * @param dateStamp The '%Y%m%d' date format.
+ * @param region The AWS region.
+ * @param serviceName The AWS service name, eg. sts.
+ * @return The signing key bytes.
+ */
+async function getSigningKey(
+  crypto: Crypto,
+  key: string,
+  dateStamp: string,
+  region: string,
+  serviceName: string
+): Promise<ArrayBuffer> {
+  const kDate = await sign(crypto, `AWS4${key}`, dateStamp);
+  const kRegion = await sign(crypto, kDate, region);
+  const kService = await sign(crypto, kRegion, serviceName);
+  const kSigning = await sign(crypto, kService, 'aws4_request');
+  return kSigning;
+}
+
+/**
+ * Generates the authentication header map needed for generating the AWS
+ * Signature Version 4 signed request.
+ *
+ * @param accessKeyId The AWS access key ID.
+ * @param secretAccessKey The AWS secret access kye.
+ * @param token The AWS token.
+ * @return The AWS authentication header map which constitutes of the following
+ *   components: amz-date, authorization header and canonical query string.
+ */
+async function generateAuthenticationHeaderMap(
+  crypto: Crypto,
+  host: string,
+  canonicalUri: string,
+  canonicalQuerystring: string,
+  method: HttpMethod,
+  region: string,
+  securityCredentials: AwsSecurityCredentials,
+  requestPayload = '',
+  additionalAmzHeaders: Headers = {}
+): Promise<AwsAuthHeaderMap> {
+  // iam.amazonaws.com host => iam service.
+  // sts.us-east-2.amazonaws.com => sts service.
+  const serviceName = host.split('.')[0];
+  const now = new Date();
+  // Format: '%Y%m%dT%H%M%SZ'.
+  const amzDate = now
+    .toISOString()
+    .replace(/[-:]/g, '')
+    .replace(/\.[0-9]+/, '');
+  // Format: '%Y%m%d'.
+  const dateStamp = now.toISOString().replace(/[-]/g, '').replace(/T.*/, '');
+
+  // Change all additional headers to be lower case.
+  const reformattedAdditionalAmzHeaders: Headers = {};
```

Comment:
This logic is definitely the most complex I think I've bumped into during this review, I believe OAuth 1.0 worked similarly, with regards to  needing to sort and sign parameters?

Reply:
I completely agree. The crypto they use is not trivial and it is not standard. You are right on the similarities with OAuth 1.0 but at least the latter is a standard and there is sufficient documentation on it. 
We had to follow the [AWS documentation](https://docs.aws.amazon.com/general/latest/gr/sigv4-calculate-signature.html) in order to generate these (step by step). They do provide reference input/outputs for each step which helped. 

I think the bulk of the logic is documented in `generateAuthenticationHeaderMap`. I provided links to the 4 steps and some comments to clarify intermediate steps. I hope this makes it easier for contributors to follow. One possibility that was floated around in the early stages was to pull in AWS SDK as a dependency. However, I think it would be too expensive for serverless and low memory environments like GCF, IOT devices, etc. The auth library is a dependency for a lot of other libraries where this feature is not used so it could be very taxing to these users.